buildSrc/src/main/java/org/elasticsearch/gradle/PublishPlugin.java

```diff
@@ -37,17 +39,38 @@
 import static org.elasticsearch.gradle.util.GradleUtils.maybeConfigure;
 
 public class PublishPlugin implements Plugin<Project> {
+    static final String MAVEN_SHADOW = "publish.maven.shadow";
+    static final String MAVEN_JAR = "publish.maven.jar";
 
     @Override
     public void apply(Project project) {
         project.getPluginManager().apply(BasePlugin.class);
-        project.getPluginManager().apply("nebula.maven-nebula-publish");
+        project.getPlugins().apply(MavenPublishPlugin.class);
         project.getPluginManager().apply(PomValidationPrecommitPlugin.class);
+        configurePublications(project);
         configureJavadocJar(project);
         configureSourcesJar(project);
         configurePomGeneration(project);
     }
 
+    private void configurePublications(Project project) {
+        ExtraPropertiesExtension extraProperties = project.getExtensions().getExtraProperties();
+        extraProperties.set(MAVEN_SHADOW, Boolean.FALSE);
+        extraProperties.set(MAVEN_JAR, Boolean.FALSE);
+
+        project.getPlugins().withType(JavaPlugin.class).configureEach(plugin -> extraProperties.set(MAVEN_JAR, Boolean.TRUE));
+        project.getPluginManager().withPlugin("com.github.johnrengelman.shadow", plugin -> extraProperties.set(MAVEN_JAR, Boolean.TRUE));
```

Comment:
Shouldn't we be mutating the `MAVEN_SHADOW` property here instead? And really, is tracking these things via properties necessary? Couldn't we simply do `configureWithShadowPlugin` here when the plugin is applied rather than setting a flag then reading it in `afterEvalutate`?

Reply:
indeed. The general problem we deal here with is applying an either or pattern. If just the java plugin is applied then  do _this_. If shadow plugin is applied do _that_. 
I looked a bit deeper into this and since how we handle the shadow plugin here we actually could simplify this without the need for the extra properties and the after evaluate hook. pushed a fix for this

Reply:
Doesn't the shadow plugin also apply the java plugin? If so can't we just order it such that it's safe to always do the "java" configuration and we just override those bits if necessary when the shadow plugin gets applied? I'm quite certain this was how it behaved before.

Reply:
No it doesn't apply it. It only reacts on the java plugin being applied. But we're good here now I think